### DISCUSSION TITLE
failed to build with SAFE_HEAP=1

### DISCUSSION THREAD

**Discussion by aharondavid** [emscripten-core/emscripten#18470] [Q&A]
Trying to build with SAFE_HEAP=1 is failed:

emcc version - 3.1.9
OS: Windows 11

**Build command line:**
`C:\emsdk\upstream\emscripten\\emcc.bat Emscripten/Debug/TEApiWrapper80_for_emscripten.o Emscripten/Debug/Main.o -ljpeg.Emscripten -lxml2 -lprjplg.Emscripten -lproj -lgdal -lgeos -lgeos_c -l3ds.Emscripten -lTerraExplorerX.Emscripten -lCore.Emscripten -l7zip.Emscripten -llibzip.Emscripten -lgisplg.Emscripten -lMptDll.Emscripten -lOpenGL.Emscripten -lPlgShell.Emscripten -lSLLogger.Emscripten -lSLLoggerLib.Emscripten -lSLStrings.Emscripten -lTerraCommon.Emscripten -lTerra.Emscripten -lTESWQ.Emscripten -lWFSPlugin.Emscripten -lsl_png.Emscripten -ldraco.Emscripten -lOpenCTM.Emscripten -lwebp.Emscripten -lMPTPlg.Emscripten -s ALLOW_TABLE_GROWTH -s WASM_BIGINT  -s EXPORTED_RUNTIME_METHODS=ccall,cwrap -s EXPORTED_FUNCTIONS=_main  -g  -gsplit-dwarf -gdwarf-5 -gpubnames --bind -s USE_PTHREADS -s ALLOW_MEMORY_GROWTH=1    -s EXIT_RUNTIME -s ASSERTIONS=2   -s FETCH   -s NO_DISABLE_EXCEPTION_CATCHING -s TEXTDECODER=0  -lidbfs.js -s PTHREAD_POOL_SIZE=60  -s STACK_OVERFLOW_CHECK=1 -s SAFE_HEAP=1 -O0 --preload-file com.skyline.terraexplorer -o C:/Projects/TerraExplorer_v8_0_0_Dev/SLSource/Emscripten/Debug/TEW/TerraExplorerApp.Emscripten.js -LC:/Projects/TerraExplorer_v8_0_0_Dev/SLSource/Emscripten/Debug/ -LC:/Projects/TerraExplorer_v8_0_0_Dev/SLSource/Emscripten/TerraExplorerApp.Emscripten/libs`

**Result:**
`emcc: error: 'C:/emsdk/upstream\bin\wasm-opt --safe-heap --strip-producers C:/Projects/TerraExplorer_v8_0_0_Dev/SLSource/Emscripten/Debug/TEW/TerraExplorerApp.Emscripten.wasm -o C:/Projects/TerraExplorer_v8_0_0_Dev/SLSource/Emscripten/Debug/TEW/TerraExplorerApp.Emscripten.wasm -g --mvp-features --enable-threads --enable-mutable-globals --enable-bulk-memory --enable-sign-ext' failed (returned 3221226505)`

Is anyone can tell what is the reason for that failing, and what is the solution ?   
`

**Comment by sbc100** [comment]
It looks like maybe wasm-opt is simply crashing.   Is this a regression?  (i.e. did it used to work with some previous version?)

Is this a large project?  (i.e. could wasm-opt simply be running out of memory?)

If this is a binaryn (wasm-opt) bug then it would be really helpful if you could attach the input file (`C:/Projects/TerraExplorer_v8_0_0_Dev/SLSource/Emscripten/Debug/TEW/TerraExplorerApp.Emscripten.wasm`).  You can preserve  this if you set EMCC_DEBUG=1 when you run the link command.

**Reply by aharondavid** [reply]
I always worked with emcc 3.1.9 ,but,  I think in some point, it worked.
Yes, it is a large Project - the wasm is about 170 M

**Reply by sbc100** [reply]
Attaching `TerraExplorerApp.Emscripten.wasm` would still be useful if you want someone to be able to debug a binaryen crash.

Another useful thing would be bisecting: https://emscripten.org/docs/contributing/developers_guide.html#bisecting

**Reply by aharondavid** [reply]
Hi,
If I understand correctly, the link above is useful when you have modules that built with a different emcc versions.  but this is not my case.
The `TerraExplorerApp.Emscripten.zip` is about 50M, so, unfortunately, I cannot attach it.

**Reply by sbc100** [reply]
Bisecting is useful if you have build tat worked with one emscripten version but fails with another.    It allows is to norrow down the cause of the regression.    In this case we know the issue started sometime after 3.1.9, and bisecting can tell us exactly when.

I suspect wasm-opt could simply be running out of memory and being killed by the OS.  Can you try linking without `-g`?  It should cause binaryen to a lot let less work.

In terms of sharing the file, if its exceeding the github limits perhaps you could share via some kind of public file sharing service?

**Comment by aharondavid** [comment]
Removing the `-g` doesn't help.

here is the wasm file:
http://dev.skylinesoft.com/up/Aharon/TerraExplorerApp.Emscripten.zip
